brut.apktool/apktool-lib/src/main/java/brut/androlib/ApkBuilder.java

```diff
@@ -141,13 +148,23 @@ private void buildManifestFile(File manifest, File manifestOriginal) throws Andr
         }
     }
 
-    private void buildSources() throws AndrolibException {
-        if (!buildSourcesRaw("classes.dex") && !buildSourcesSmali("smali", "classes.dex")) {
-            LOGGER.warning("Could not find sources");
-        }
+    private void scheduleBuildSources() {
+        Runnable r = () -> {
+            try {
+                if (mBuildError.get() != null) {
+                    return;
+                }
+                if (!buildSourcesRaw("classes.dex") && !buildSourcesSmali("smali", "classes.dex")) {
+                    LOGGER.warning("Could not find sources");
+                }
+            } catch (AndrolibException e) {
+                mBuildError.compareAndSet(null, e);
+            }
+        };
+        mWorker.submit(r);
     }
 
-    private void buildNonDefaultSources() throws AndrolibException {
+    private void scheduleBuildNonDefaultSources() throws AndrolibException {
```

Comment:
Should we extract the common logic that seems to be the scheduling of the disassembly of dexs into its own method? It seems the single one (classes.dex) and iterating multi-dex is setting up the same run-able to submit.